server/channels/app/import_functions.go

```diff
@@ -1296,58 +1348,80 @@ func (a *App) importAttachment(rctx request.CTX, data *imports.AttachmentImportD
 		}
-		rctx.Logger().Info("Preparing file upload from file system", fields...)
+		fileSize = info.Size()
+
+		rctx.Logger().Info("Preparing file upload from file system", mlog.String("file_name", name), mlog.Int("file_size", info.Size()))
 	}
 
 	timestamp := utils.TimeFromMillis(post.CreateAt)
 
-	fileData, err := io.ReadAll(file)
-	if err != nil {
-		return nil, model.NewAppError("BulkImport", "app.import.attachment.read_file_data.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest)
-	}
-
 	// Go over existing files in the post and see if there already exists a file with the same name, size and hash. If so - skip it
 	if post.Id != "" {
-		oldFiles, err := a.getFileInfosForPostIgnoreCloudLimit(rctx, post.Id, true, false)
+		oldFiles, err := a.Srv().Store().FileInfo().GetForPost(post.Id, true, false, true)
 		if err != nil {
 			return nil, model.NewAppError("BulkImport", "app.import.attachment.file_upload.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest)
 		}
 		for _, oldFile := range oldFiles {
-			if oldFile.Name != path.Base(name) || oldFile.Size != int64(len(fileData)) {
+			if oldFile.Name != path.Base(name) || oldFile.Size != fileSize {
 				continue
 			}
 
-			// check sha1
-			newHash := sha1.Sum(fileData)
-			oldFileData, err := a.getFileIgnoreCloudLimit(rctx, oldFile.Id)
-			if err != nil {
+			oldFileReader, appErr := a.FileReader(oldFile.Path)
+			if appErr != nil {
 				return nil, model.NewAppError("BulkImport", "app.import.attachment.file_upload.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest)
 			}
-			oldHash := sha1.Sum(oldFileData)
+			defer oldFileReader.Close()
 
-			if bytes.Equal(oldHash[:], newHash[:]) {
-				rctx.Logger().Info("Skipping uploading of file because name already exists", mlog.String("file_name", name))
+			if ok, err := compareFilesContent(oldFileReader, file, 0); err != nil {
+				rctx.Logger().Error("Failed to compare files content", mlog.String("file_name", name), mlog.Err(err))
+			} else if ok {
+				rctx.Logger().Info("Skipping uploading of file because name already exists and content matches", mlog.String("file_name", name))
 				return oldFile, nil
 			}
+
+			rctx.Logger().Info("File contents don't match, will re-upload", mlog.String("file_name", name))
+
+			// Since compareFilesContent needs to read the whole file we need to
+			// either seek back (local file) or re-open it (zip file).
+			if f, ok := file.(*os.File); ok {
+				rctx.Logger().Info("File is *os.File, can seek")
+				if _, err := f.Seek(0, io.SeekStart); err != nil {
+					return nil, model.NewAppError("BulkImport", "app.import.attachment.seek_file.error", map[string]any{"FilePath": *data.Path}, "", http.StatusBadRequest).Wrap(err)
+				}
+			} else if data.Data != nil {
+				rctx.Logger().Info("File is from ZIP, can't seek, opening again")
```

Comment:
Can we add some more context here? e.g. which file? line number? etc.

Reply:
Makes sense :+1: